README.md
# The OpenBCI Web Bluetooth Driver - !!WARNING!! STILL IN PRE-ALPHA PHASE

<p align="center">
  <img alt="banner" src="/images/ganglion_head_shot.jpg/" width="600">
</p>
<p align="center" href="">
  Provide a stable cross platform web bluetooth driver for the OpenBCI Ganglion
</p>


## Welcome!

First and foremost, Welcome! :tada: Willkommen! :confetti_ball: Bienvenue! :balloon::balloon::balloon:

Thank you for visiting the OpenBCI Ganglion Web Bluetooth repository.

This document (the README file) is a hub to give you some information about the project. Jump straight to one of the sections below, or just scroll down to find out more.

* [What are we doing? (And why?)](#what-are-we-doing)
* [Who are we?](#who-are-we)
* [What do we need?](#what-do-we-need)
* [How can you get involved?](#get-involved)
* [Get in touch](#contact-us)
* [Find out more](#find-out-more)
* [Installing](#installing)

## What are we doing?

### The problem

* OpenBCI Ganglion NodeJS must rely on [`noble`](https://github.com/sandeepmistry/noble) for bluetooth control.
* Some linux and macOS users drop a lot of packets using the current drivers
* There is no cross platform driver for Ganglion right now
* People must use NodeJS always
* Ganglion data is compressed and in a very raw form, takes programmers a while to understand

So, it's clear there are some issues that need to be fixed!

### The solution

The OpenBCI Ganglion Web Bluetooth driver will:

* Will leverage web bluetooth which is supported by a massive community
* Hopfully solve the dropped packet problem on mac and linux
* Avoid windows users from having to use a dongle
* Be fun and easy to use
* Take the hard work of decompressing the raw data stream for programmers

Using the OpenBCI Ganglion Web Bluetooth driver allows you, the user, to quickly

## Who are we?

Mainly, we are community of programmers who love brainwaves on the browser. The original code writer was [Uri Shaked][link_uri_shaked]'s and we hard forked his [muse-js](https://github.com/urish/muse-lsl) library. AJ Keller overhauled and applied his work from [OpenBCI_Ganglion_NodeJS](github.com/OpenBCI/OpenBCI_Ganglion_NodeJS).

## What do we need?

**You**! In whatever way you can help.

We need expertise in programming, user experience, software sustainability, documentation and technical writing and project management.

We'd love your feedback along the way.

Our primary goal is to brovide a stable cross platform web bluetooth driver for the OpenBCI Ganglion, and we're excited to support the professional development of any and all of our contributors. If you're looking to learn to code, try out working collaboratively, or translate you skills to the digital domain, we're here to help.

## Get involved

If you think you can help in any of the areas listed above (and we bet you can) or in any of the many areas that we haven't yet thought of (and here we're *sure* you can) then please check out our [contributors' guidelines](CONTRIBUTING.md) and our [roadmap](ROADMAP.md).

Please note that it's very important to us that we maintain a positive and supportive environment for everyone who wants to participate. When you join us we ask that you follow our [code of conduct](CODE_OF_CONDUCT.md) in all interactions both on and offline.


## Contact us

If you want to report a problem or suggest an enhancement we'd love for you to [open an issue](../../issues) at this github repository because then we can get right on it. But you can also contact [AJ][link_aj_keller] by email (pushtheworldllc AT gmail DOT com) or on [twitter](https://twitter.com/aj-ptw).

## Find out more

You might be interested in:

* Purchase a [Cyton][link_shop_cyton] | [Ganglion][link_shop_ganglion] | [WiFi Shield][link_shop_wifi_shield] from [OpenBCI][link_openbci]

And of course, you'll want to know our:

* [Contributors' guidelines](CONTRIBUTING.md)
* [Roadmap](ROADMAP.md)

## Thank you

Thank you so much (Danke schön! Merci beaucoup!) for visiting the project and we do hope that you'll join us on this amazing journey to provide a stable cross platform web bluetooth driver for the OpenBCI Ganglion.

## Running the demo app

    yarn
    yarn start

and then open http://localhost:4445/

## Usage example

```javascript

import { GanglionClient } from 'openbci-ganglion-js';

async function main() {
  let client = new GanglionClient();
  await client.connect();
  await client.start();
  client.eegReadings.subscribe(reading => {
    console.log(reading);
  });
  client.accelerometerData.subscribe(acceleration => {
    console.log(acceleration);
  });
}

main();
```

## <a name="license"></a> License:

MIT

[link_aj_keller]: https://github.com/aj-ptw
[link_shop_wifi_shield]: https://shop.openbci.com/collections/frontpage/products/wifi-shield?variant=44534009550
[link_shop_ganglion]: https://shop.openbci.com/collections/frontpage/products/pre-order-ganglion-board
[link_shop_cyton]: https://shop.openbci.com/collections/frontpage/products/cyton-biosensing-board-8-channel
[link_shop_cyton_daisy]: https://shop.openbci.com/collections/frontpage/products/cyton-daisy-biosensing-boards-16-channel
[link_ptw]: https://www.pushtheworldllc.com
[link_openbci]: http://www.openbci.com
[link_gui_widget_tutorial]: http://docs.openbci.com/Tutorials/15-Custom_Widgets
[link_gui_run_from_processing]: http://docs.openbci.com/OpenBCI%20Software/01-OpenBCI_GUI#the-openbci-gui-running-the-openbci-gui-from-the-processing-ide
[link_uri_shaked]: https://github.com/urish

CHANGELOG.md
There have been no releases because this repo is unstable.


src/lib/ganglion-utils.spec.ts
import { encodeCommand } from './ganglion-utils';

import { TextEncoder, TextDecoder } from 'text-encoding'; // polyfill

declare var global: any;
global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;

describe('encodeCommand', () => {
    it('should correctly encode the given command as a Uint8Array', () => {
        expect(encodeCommand('v1')).toEqual(new Uint8Array([3, 118, 49, 10]));
    })
});


src/muse.ts
import { Observable } from 'rxjs/Observable';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import 'rxjs/add/observable/merge';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/first';
import 'rxjs/add/operator/share';
import 'rxjs/add/operator/take';
import 'rxjs/add/operator/toPromise';

import { EEGReading, TelemetryData, AccelerometerData, GyroscopeData, XYZ, MuseControlResponse, MuseDeviceInfo } from './lib/muse-interfaces';
import { parseControl, decodeEEGSamples, parseTelemetry, parseAccelerometer, parseGyroscope } from './lib/muse-parse';
import { encodeCommand, decodeResponse, observableCharacteristic } from './lib/muse-utils';

export { zipSamples, EEGSample } from './lib/zip-samples';
export { EEGReading, TelemetryData, AccelerometerData, GyroscopeData, XYZ, MuseControlResponse };

export const SIMBLEE_SERVICE = 0xfe84;
/** Simblee */
const CONTROL_CHARACTERISTIC = '2d30c083f39f4ce6923f3484ea480596';
const EEG_CHARACTERISTIC = '2d30c082f39f4ce6923f3484ea480596';

// These names match the characteristics defined in EEG_CHARACTERISTICS above
export const channelNames = [
    'TP9',
    'AF7',
    'AF8',
    'TP10',
    'AUX'
];

export class MuseClient {
    private gatt: BluetoothRemoteGATTServer | null = null;
    private controlChar: BluetoothRemoteGATTCharacteristic;
    private eegCharacteristics: BluetoothRemoteGATTCharacteristic[];

    public enableAux = false;
    public deviceName: string | null = '';
    public connectionStatus = new BehaviorSubject<boolean>(false);
    public rawControlData: Observable<string>;
    public controlResponses: Observable<MuseControlResponse>;
    public accelerometerData: Observable<AccelerometerData>;
    public eegReadings: Observable<EEGReading>;

    async connect(gatt?: BluetoothRemoteGATTServer) {
        if (gatt) {
            this.gatt = gatt;
        } else {
            const device = await navigator.bluetooth.requestDevice({
                filters: [{ services: [SIMBLEE_SERVICE] }]
            });
            this.gatt = await device.gatt!.connect();
        }
        this.deviceName = this.gatt.device.name || null;

        const service = await this.gatt.getPrimaryService(SIMBLEE_SERVICE);
        Observable.fromEvent<void>(this.gatt.device, 'gattserverdisconnected').first().subscribe(() => {
            this.gatt = null;
            this.connectionStatus.next(false);
        });

        // Control
        this.controlChar = await service.getCharacteristic(CONTROL_CHARACTERISTIC);
        this.rawControlData = (await observableCharacteristic(this.controlChar))
            .map(data => decodeResponse(new Uint8Array(data.buffer)))
            .share();
        this.controlResponses = parseControl(this.rawControlData);

        // Battery
        const telemetryCharacteristic = await service.getCharacteristic(TELEMETRY_CHARACTERISTIC);
        this.telemetryData = (await observableCharacteristic(telemetryCharacteristic))
            .map(parseTelemetry);

        // Gyroscope
        const gyroscopeCharacteristic = await service.getCharacteristic(GYROSCOPE_CHARACTERISTIC);
        this.gyroscopeData = (await observableCharacteristic(gyroscopeCharacteristic))
            .map(parseGyroscope);

        // Accelerometer
        const eegCharacteristic = await service.getCharacteristic(EEG_CHARACTERISTIC);
        this.eegData = (await observableCharacteristic(eegCharacteristic))
            .map(parseEEGData);

        // EEG
        this.eegCharacteristics = [];
        const eegObservables = [];
        const channelCount = this.enableAux ? EEG_CHARACTERISTICS.length : 4;
        for (let index = 0; index < channelCount; index++) {
            let characteristicId = EEG_CHARACTERISTICS[index];
            const eegChar = await service.getCharacteristic(characteristicId);
            eegObservables.push(
                (await observableCharacteristic(eegChar)).map(data => {
                    return {
                        electrode: index,
                        timestamp: data.getUint16(0),
                        samples: decodeEEGSamples(new Uint8Array(data.buffer).subarray(2))
                    };
                }));
            this.eegCharacteristics.push(eegChar);
        }
        this.eegReadings = Observable.merge(...eegObservables);
        this.connectionStatus.next(true);
    }

    async sendCommand(cmd: string) {
        await this.controlChar.writeValue((encodeCommand(cmd)));
    }

    async start() {
        // Subscribe to egg characteristics
        await this.pause();
        // Select preset number 20
        await this.controlChar.writeValue(encodeCommand('p20'));
        await this.controlChar.writeValue(encodeCommand('s'));
        await this.resume();
    }

    async pause() {
        await this.sendCommand('h');
    }

    async resume() {
        await this.sendCommand('d');
    }

    async deviceInfo() {
        const resultListener = this.controlResponses.filter(r => !!r.fw).take(1).toPromise();
        await this.sendCommand('v1');
        return resultListener as Promise<MuseDeviceInfo>;
    }

    disconnect() {
        if (this.gatt) {
            this.gatt.disconnect();
            this.connectionStatus.next(false);
        }
    }
}


src/ganglion.ts
import { Observable } from 'rxjs/Observable';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import 'rxjs/add/observable/merge';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/first';
import 'rxjs/add/operator/share';
import 'rxjs/add/operator/take';
import 'rxjs/add/operator/toPromise';

import { EEGReading, AccelerometerData, GanglionControlResponse, GanglionDeviceInfo } from './lib/ganglion-interfaces';
import { parseControl, decodeEEGSamples } from './lib/ganglion-parse';
import { encodeCommand, decodeResponse, observableCharacteristic } from './lib/ganglion-utils';

export { zipSamples, EEGSample } from './lib/zip-samples';
export { EEGReading, AccelerometerData, GanglionControlResponse };

export const SIMBLEE_SERVICE = 0xfe84;
/** Simblee */
const CONTROL_CHARACTERISTIC = '2d30c083f39f4ce6923f3484ea480596';
const EEG_CHARACTERISTIC = '2d30c082f39f4ce6923f3484ea480596';

// These names match the characteristics defined in EEG_CHARACTERISTICS above
export const channelNames = [
    '1',
    '2',
    '3',
    '4'
];

export class GanglionClient {
    private gatt: BluetoothRemoteGATTServer | null = null;
    private controlChar: BluetoothRemoteGATTCharacteristic;
    private eegCharacteristics: BluetoothRemoteGATTCharacteristic[];

    public enableAux = false;
    public deviceName: string | null = '';
    public connectionStatus = new BehaviorSubject<boolean>(false);
    public rawControlData: Observable<string>;
    public controlResponses: Observable<GanglionControlResponse>;
    public eegReadings: Observable<EEGReading>;

    async connect(gatt?: BluetoothRemoteGATTServer) {
        if (gatt) {
            this.gatt = gatt;
        } else {
            const device = await navigator.bluetooth.requestDevice({
                filters: [{ services: [SIMBLEE_SERVICE] }]
            });
            this.gatt = await device.gatt!.connect();
        }
        this.deviceName = this.gatt.device.name || null;

        const service = await this.gatt.getPrimaryService(SIMBLEE_SERVICE);
        Observable.fromEvent<void>(this.gatt.device, 'gattserverdisconnected').first().subscribe(() => {
            this.gatt = null;
            this.connectionStatus.next(false);
        });

        // Control
        this.controlChar = await service.getCharacteristic(CONTROL_CHARACTERISTIC);
        this.rawControlData = (await observableCharacteristic(this.controlChar))
            .map(data => decodeResponse(new Uint8Array(data.buffer)))
            .share();
        this.controlResponses = parseControl(this.rawControlData);

        // EEG
        const decompressedSamples = new Array(3);
        for (let i = 0; i < 3; i++) {
            decompressedSamples[i] = [0, 0, 0, 0];
        }
        const accelArray = [0, 0, 0];
        const eegCharacteristic = await service.getCharacteristic(EEG_CHARACTERISTIC);
        this.eegReadings = (await observableCharacteristic(eegCharacteristic))
            .map(data => {
                return decodeEEGSamples(data, decompressedSamples, accelArray);
            });

        this.connectionStatus.next(true);
    }

    async sendCommand(cmd: string) {
        await this.controlChar.writeValue((encodeCommand(cmd)));
    }

    async start() {
        // Subscribe to egg characteristics
        await this.pause();
        await this.sendCommand('b');
        await this.resume();
    }

    async pause() {
        await this.sendCommand('s');
    }

    async resume() {
        await this.sendCommand('b');
    }

    async deviceInfo() {
        const resultListener = this.controlResponses.filter(r => !!r.fw).take(1).toPromise();
        await this.sendCommand('v');
        return resultListener as Promise<GanglionDeviceInfo>;
    }

    disconnect() {
        if (this.gatt) {
            this.gatt.disconnect();
            this.connectionStatus.next(false);
        }
    }
}


demo/src/main.ts
import { GanglionClient, EEGReading, channelNames } from './../../src/ganglion';

(window as any).connect = async () => {
    let graphTitles = Array.from(document.querySelectorAll('.electrode-item h3'));
    let canvases = Array.from(document.querySelectorAll('.electrode-item canvas')) as HTMLCanvasElement[];
    let canvasCtx = canvases.map(canvas => canvas.getContext('2d'));

    graphTitles.forEach((item, index) => {
        item.textContent = channelNames[index];
    });

    function plot(reading: EEGReading) {
        const canvas = canvases[reading.electrode];
        const context = canvasCtx[reading.electrode];
        if (!context) {
            return;
        }
        const width = canvas.width / 12.0;
        const height = canvas.height / 2.0;
        context.fillStyle = 'green';
        context.clearRect(0, 0, canvas.width, canvas.height);

        for (let i = 0; i < reading.samples.length; i++) {
            const sample = reading.samples[i] / 15.;
            if (sample > 0) {
                context.fillRect(i * 25, height - sample, width, sample);
            } else {
                context.fillRect(i * 25, height, width, -sample);
            }
        }
    }

    let client = new GanglionClient();
    client.connectionStatus.subscribe(status => {
        console.log(status ? 'Connected!' : 'Disconnected')
    });

    try {
        // client.enableAux = true;
        await client.connect();
        await client.start();
        document.getElementById('headset-name')!.innerText = client.deviceName;
        client.eegReadings.subscribe(reading => {
            plot(reading);
        });
        // client.telemetryData.subscribe(reading => {
        //     document.getElementById('temperature')!.innerText = reading.temperature.toString() + '℃';
        //     document.getElementById('batteryLevel')!.innerText = reading.batteryLevel.toFixed(2) + '%';
        // });
        // client.accelerometerData.subscribe(accel => {
        //     const normalize = (v: number) => (v / 16384.).toFixed(2) + 'g';
        //     document.getElementById('accelerometer-x')!.innerText = normalize(accel.samples[2].x);
        //     document.getElementById('accelerometer-y')!.innerText = normalize(accel.samples[2].y);
        //     document.getElementById('accelerometer-z')!.innerText = normalize(accel.samples[2].z);
        // });
        // await client.deviceInfo().then(deviceInfo => {
        //     document.getElementById('hardware-version')!.innerText = deviceInfo.hw;
        //     document.getElementById('firmware-version')!.innerText = deviceInfo.fw;
        // });
    } catch (err) {
        console.error('Connection failed', err);
    }
};


src/lib/zip-samples.ts
import { Observable } from 'rxjs/Observable';
import { EEGReading } from './ganglion-interfaces';

import 'rxjs/add/observable/from';
import 'rxjs/add/operator/mergeMap';
import 'rxjs/add/operator/concat';

export interface EEGSample {
    timestamp: number;
    data: number[];
};

export function zipSamples(eegReadings: Observable<EEGReading>): Observable<EEGSample> {
    const buffer: EEGReading[] = [];
    let lastTimestamp: number | null = null;
    return eegReadings
        .mergeMap<EEGReading, EEGReading[]>(reading => {
            if (reading.timestamp !== lastTimestamp) {
                lastTimestamp = reading.timestamp;
                if (buffer.length) {
                    const result = Observable.from([[...buffer]]);
                    buffer.splice(0, buffer.length, reading);
                    return result;
                }
            }
            buffer.push(reading);
            return Observable.from([]);
        })
        .concat(Observable.from([buffer]))
        .mergeMap((readings: EEGReading[]) => {
            const result = readings[0].samples.map((x, index) => {
                const data = [NaN, NaN, NaN, NaN, NaN];
                for (let reading of readings) {
                    data[reading.electrode] = reading.samples[index];
                }
                return {
                    timestamp: readings[0].timestamp,
                    data
                };
            });
            return Observable.from(result);
        });
}


src/lib/ganglion-parse.ts
import { Observable } from 'rxjs/Observable';

import 'rxjs/add/operator/concatMap';
import 'rxjs/add/operator/scan';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/map';

import { EEGReading, AccelerometerData } from './ganglion-interfaces';
import { k } from './ganglion-utils';

export function parseControl(controlData: Observable<string>) {
    return controlData
        .concatMap(data => data.split(''))
        .scan((acc, value) => {
            if (acc.indexOf('}') >= 0) {
                return value;
            } else {
                return acc + value;
            }
        }, '')
        .filter(value => value.indexOf('}') >= 0)
        .map(value => JSON.parse(value));
}

export function decodeUnsigned12BitData(samples: Uint8Array) {
    const samples12Bit = [];
    for (let i = 0; i < samples.length; i++) {
        if (i % 3 === 0) {
            samples12Bit.push(samples[i] << 4 | samples[i + 1] >> 4);
        } else {
            samples12Bit.push((samples[i] & 0xf) << 8 | samples[i + 1]);
            i++;
        }
    }
    return samples12Bit;
}

export function decodeEEGSamples(data: Uint8Array, decompressedSamples: Array<Array<number>>, accelArray: Array<number>) {
    return _processRouteSampleData(data, decompressedSamples, accelArray);
    // ATTENTION: FOR LATER WE MUST PARSE RAW DATA BASED ON BYTE ID
    // let byteId = data[0];
    // if (byteId <= k.OBCIGanglionByteId19Bit.max) {
    //     _processRouteSampleData(data);
    // } else {
    //     switch (byteId) {
    //         case k.OBCIGanglionByteIdMultiPacket:
    //             this._processMultiBytePacket(data);
    //             break;
    //         case k.OBCIGanglionByteIdMultiPacketStop:
    //             this._processMultiBytePacketStop(data);
    //             break;
    //         case k.OBCIGanglionByteIdImpedanceChannel1:
    //         case k.OBCIGanglionByteIdImpedanceChannel2:
    //         case k.OBCIGanglionByteIdImpedanceChannel3:
    //         case k.OBCIGanglionByteIdImpedanceChannel4:
    //         case k.OBCIGanglionByteIdImpedanceChannelReference:
    //             this._processImpedanceData(data);
    //             break;
    //         default:
    //             this._processOtherData(data);
    //     }
    // }
}

// //////// //
// PRIVATES //
// //////// //
/**
 * Builds a sample object from an array and sample number.
 * @param decompressedSamples {Array} - An array that of three
 *  samples to compute and store deltas through each cycle
 * @return {{sampleNumber: *}}
 * @private
 */
function _buildSampleCompressed (decompressedSamples: Array<Array<number>>) {
    let out = [];
    for (let i = 0; i < k.OBCINumberOfChannelsGanglion; i++) {
        out.push({
            electrode: i,
            timestamp: Date.now(),
            samples: [
                decompressedSamples[1][i]  * k.OBCIGanglionScaleFactorPerCountVolts,
                decompressedSamples[2][i]  * k.OBCIGanglionScaleFactorPerCountVolts
            ]
        })
    }
    return out;
}

/**
 * Builds a sample object from an array and sample number.
 * @param decompressedSamples {Array} - An array that of three
 *  samples to compute and store deltas through each cycle
 * @return {{sampleNumber: *}}
 * @private
 */
function _buildSampleDecompressed (decompressedSamples: Array<Array<number>>) {
    let out = [];
    for (let i = 0; i < k.OBCINumberOfChannelsGanglion; i++) {
        out.push({
            electrode: i,
            timestamp: Date.now(),
            samples: [
                decompressedSamples[0][i]  * k.OBCIGanglionScaleFactorPerCountVolts
            ]
        })
    }
    return out;
}

/**
 * Utilize `receivedDeltas` to get actual count values.
 * @param receivedDeltas {Array} - An array of deltas
 *  of shape 2x4 (2 samples per packet and 4 channels per sample.)
 * @param decompressedSamples {Array} - An array that of three
 *  samples to compute and store deltas through each cycle
 * @private
 */
export function _decompressSamples(receivedDeltas: Array<Array<number>>, decompressedSamples: Array<Array<number>>) {
    // add the delta to the previous value
    for (let i = 1; i < 3; i++) {
        for (let j = 0; j < 4; j++) {
            decompressedSamples[i][j] = decompressedSamples[i - 1][j] - receivedDeltas[i - 1][j];
        }
    }
}

/**
 * Process an compressed packet of data.
 * @param data {Uint8Array}
 *  Data packet buffer from noble.
 * @param decompressedSamples {Array} - An array that of three
 *  samples to compute and store deltas through each cycle
*  @param accelArray {Array} - An array to hold casual inter packet data
 * @private
 */
export function _processCompressedData(data: Uint8Array, decompressedSamples: Array<Array<number>>, accelArray: Array<number>) {
    // Save the packet counter
    let packetCounter = data[0];
    let eegReadings;
    // Decompress the buffer into array
    if (packetCounter <= k.OBCIGanglionByteId18Bit.max) {
        _decompressSamples(decompressDeltas18Bit(data.slice(k.OBCIGanglionPacket18Bit.dataStart, k.OBCIGanglionPacket18Bit.dataStop)), decompressedSamples);
        eegReadings = _buildSampleCompressed(decompressedSamples);

        switch (packetCounter % 10) {
            case k.OBCIGanglionAccelAxisX:
                accelArray[0] = data[k.OBCIGanglionPacket18Bit.auxByte - 1] * k.OBCIGanglionAccelScaleFactor;
                break;
            case k.OBCIGanglionAccelAxisY:
                accelArray[1] = data[k.OBCIGanglionPacket18Bit.auxByte - 1] * k.OBCIGanglionAccelScaleFactor;
                break;
            case k.OBCIGanglionAccelAxisZ:
                accelArray[2] = data[k.OBCIGanglionPacket18Bit.auxByte - 1] * k.OBCIGanglionAccelScaleFactor;
                break;
            default:
                break;
        }
    } else {
        _decompressSamples(decompressDeltas19Bit(data.slice(k.OBCIGanglionPacket19Bit.dataStart, k.OBCIGanglionPacket19Bit.dataStop)), decompressedSamples);
        eegReadings = _buildSampleCompressed(decompressedSamples);
    }

    // Rotate the 0 position for next time
    for (let i = 0; i < k.OBCINumberOfChannelsGanglion; i++) {
        decompressedSamples[0][i] = decompressedSamples[2][i];
    }

    return eegReadings;
}

export function _processRouteSampleData(data: Uint8Array, decompressedSamples: Array<Array<number>>, accelArray: Array<number>) {
    if (data[0] === k.OBCIGanglionByteIdUncompressed) {
        _processUncompressedData(data, decompressedSamples);
    } else {
        _processCompressedData(data, decompressedSamples, accelArray);
    }
}

/**
 * Process an uncompressed packet of data.
 * @param data {Buffer}
 *  Data packet buffer from noble.
 * @param decompressedSamples {Array} - An array that of three
 *  samples to compute and store deltas through each cycle
 * @private
 */
export function _processUncompressedData(data: Uint8Array, decompressedSamples: Array<Array<number>>) {
    let start = 1;

    for (let i = 0; i < 4; i++) {
        decompressedSamples[0][i] = interpret24bitAsInt32(data, start);  // seed the decompressor
        start += 3;
    }

    return _buildSampleDecompressed(decompressedSamples);
}

/**
 * Converts a special ganglion 18 bit compressed number
 *  The compressions uses the LSB, bit 1, as the signed bit, instead of using
 *  the MSB. Therefore you must not look to the MSB for a sign extension, one
 *  must look to the LSB, and the same rules applies, if it's a 1, then it's a
 *  negative and if it's 0 then it's a positive number.
 * @param threeByteBuffer {Uint8Array}
 *  A 3-byte buffer with only 18 bits of actual data.
 * @return {number} A signed integer.
 */
export function convert18bitAsInt32 (threeByteBuffer: Uint8Array) {
  let prefix = 0;

  if ((threeByteBuffer[2] & 0x01) > 0) {
    // console.log('\t\tNegative number')
    prefix = 0b11111111111111;
  }

  return (prefix << 18) | (threeByteBuffer[0] << 16) | (threeByteBuffer[1] << 8) | threeByteBuffer[2];
}

/**
 * Converts a special ganglion 19 bit compressed number
 *  The compressions uses the LSB, bit 1, as the signed bit, instead of using
 *  the MSB. Therefore you must not look to the MSB for a sign extension, one
 *  must look to the LSB, and the same rules applies, if it's a 1, then it's a
 *  negative and if it's 0 then it's a positive number.
 * @param threeByteBuffer {Uint8Array}
 *  A 3-byte buffer with only 19 bits of actual data.
 * @return {number} A signed integer.
 */
export function convert19bitAsInt32 (threeByteBuffer: Uint8Array) {
  let prefix = 0;

  if ((threeByteBuffer[2] & 0x01) > 0) {
    // console.log('\t\tNegative number')
    prefix = 0b1111111111111;
  }

  return (prefix << 19) | (threeByteBuffer[0] << 16) | (threeByteBuffer[1] << 8) | threeByteBuffer[2];
}

/**
 * Called to when a compressed packet is received.
 * @param buffer {Uint8Array} Just the data portion of the sample. So 18 bytes.
 * @return {Array} - An array of deltas of shape 2x4 (2 samples per packet
 *  and 4 channels per sample.)
 */
export function decompressDeltas18Bit (buffer: Uint8Array) {
  let D = new Array(k.OBCIGanglionSamplesPerPacket); // 2 Ganglion Samples Per Packet
  D[0] = [0, 0, 0, 0];
  D[1] = [0, 0, 0, 0];

  let receivedDeltas = [];
  for (let i = 0; i < k.OBCIGanglionSamplesPerPacket; i++) {
    receivedDeltas.push([0, 0, 0, 0]);
  }

  let miniBuf;

  // Sample 1 - Channel 1
  miniBuf = new Uint8Array(
    [
      (buffer[0] >> 6),
      ((buffer[0] & 0x3F) << 2) | (buffer[1] >> 6),
      ((buffer[1] & 0x3F) << 2) | (buffer[2] >> 6)
    ]
  );
  receivedDeltas[0][0] = convert18bitAsInt32(miniBuf);

  // Sample 1 - Channel 2
  miniBuf = new Uint8Array(
    [
      (buffer[2] & 0x3F) >> 4,
      (buffer[2] << 4) | (buffer[3] >> 4),
      (buffer[3] << 4) | (buffer[4] >> 4)
    ]);
  receivedDeltas[0][1] = convert18bitAsInt32(miniBuf);

  // Sample 1 - Channel 3
  miniBuf = new Uint8Array(
    [
      (buffer[4] & 0x0F) >> 2,
      (buffer[4] << 6) | (buffer[5] >> 2),
      (buffer[5] << 6) | (buffer[6] >> 2)
    ]);
  receivedDeltas[0][2] = convert18bitAsInt32(miniBuf);

  // Sample 1 - Channel 4
  miniBuf = new Uint8Array(
    [
      (buffer[6] & 0x03),
      buffer[7],
      buffer[8]
    ]);
  receivedDeltas[0][3] = convert18bitAsInt32(miniBuf);

  // Sample 2 - Channel 1
  miniBuf = new Uint8Array(
    [
      (buffer[9] >> 6),
      ((buffer[9] & 0x3F) << 2) | (buffer[10] >> 6),
      ((buffer[10] & 0x3F) << 2) | (buffer[11] >> 6)
    ]);
  receivedDeltas[1][0] = convert18bitAsInt32(miniBuf);

  // Sample 2 - Channel 2
  miniBuf = new Uint8Array(
    [
      (buffer[11] & 0x3F) >> 4,
      (buffer[11] << 4) | (buffer[12] >> 4),
      (buffer[12] << 4) | (buffer[13] >> 4)
    ]);
  receivedDeltas[1][1] = convert18bitAsInt32(miniBuf);

  // Sample 2 - Channel 3
  miniBuf = new Uint8Array(
    [
      (buffer[13] & 0x0F) >> 2,
      (buffer[13] << 6) | (buffer[14] >> 2),
      (buffer[14] << 6) | (buffer[15] >> 2)
    ]);
  receivedDeltas[1][2] = convert18bitAsInt32(miniBuf);

  // Sample 2 - Channel 4
  miniBuf = new Uint8Array([(buffer[15] & 0x03), buffer[16], buffer[17]]);
  receivedDeltas[1][3] = convert18bitAsInt32(miniBuf);

  return receivedDeltas;
}

/**
 * Called to when a compressed packet is received.
 * @param buffer {Uint8Array} Just the data portion of the sample. So 19 bytes.
 * @return {Array} - An array of deltas of shape 2x4 (2 samples per packet
 *  and 4 channels per sample.)
 * @private
 */
export function decompressDeltas19Bit (buffer: Uint8Array) {
  let D = new Array(k.OBCIGanglionSamplesPerPacket); // 2
  D[0] = [0, 0, 0, 0];
  D[1] = [0, 0, 0, 0];

  let receivedDeltas = [];
  for (let i = 0; i < k.OBCIGanglionSamplesPerPacket; i++) {
    receivedDeltas.push([0, 0, 0, 0]);
  }

  let miniBuf;

  // Sample 1 - Channel 1
  miniBuf = new Uint8Array(
    [
      (buffer[0] >> 5),
      ((buffer[0] & 0x1F) << 3) | (buffer[1] >> 5),
      ((buffer[1] & 0x1F) << 3) | (buffer[2] >> 5)
    ]
  );
  receivedDeltas[0][0] = convert19bitAsInt32(miniBuf);

  // Sample 1 - Channel 2
  miniBuf = new Uint8Array(
    [
      (buffer[2] & 0x1F) >> 2,
      (buffer[2] << 6) | (buffer[3] >> 2),
      (buffer[3] << 6) | (buffer[4] >> 2)
    ]);
  receivedDeltas[0][1] = convert19bitAsInt32(miniBuf);

  // Sample 1 - Channel 3
  miniBuf = new Uint8Array(
    [
      ((buffer[4] & 0x03) << 1) | (buffer[5] >> 7),
      ((buffer[5] & 0x7F) << 1) | (buffer[6] >> 7),
      ((buffer[6] & 0x7F) << 1) | (buffer[7] >> 7)
    ]);
  receivedDeltas[0][2] = convert19bitAsInt32(miniBuf);

  // Sample 1 - Channel 4
  miniBuf = new Uint8Array(
    [
      ((buffer[7] & 0x7F) >> 4),
      ((buffer[7] & 0x0F) << 4) | (buffer[8] >> 4),
      ((buffer[8] & 0x0F) << 4) | (buffer[9] >> 4)
    ]);
  receivedDeltas[0][3] = convert19bitAsInt32(miniBuf);

  // Sample 2 - Channel 1
  miniBuf = new Uint8Array(
    [
      ((buffer[9] & 0x0F) >> 1),
      (buffer[9] << 7) | (buffer[10] >> 1),
      (buffer[10] << 7) | (buffer[11] >> 1)
    ]);
  receivedDeltas[1][0] = convert19bitAsInt32(miniBuf);

  // Sample 2 - Channel 2
  miniBuf = new Uint8Array(
    [
      ((buffer[11] & 0x01) << 2) | (buffer[12] >> 6),
      (buffer[12] << 2) | (buffer[13] >> 6),
      (buffer[13] << 2) | (buffer[14] >> 6)
    ]);
  receivedDeltas[1][1] = convert19bitAsInt32(miniBuf);

  // Sample 2 - Channel 3
  miniBuf = new Uint8Array(
    [
      ((buffer[14] & 0x38) >> 3),
      ((buffer[14] & 0x07) << 5) | ((buffer[15] & 0xF8) >> 3),
      ((buffer[15] & 0x07) << 5) | ((buffer[16] & 0xF8) >> 3)
    ]);
  receivedDeltas[1][2] = convert19bitAsInt32(miniBuf);

  // Sample 2 - Channel 4
  miniBuf = new Uint8Array([(buffer[16] & 0x07), buffer[17], buffer[18]]);
  receivedDeltas[1][3] = convert19bitAsInt32(miniBuf);

  return receivedDeltas;
}

function interpret24bitAsInt32 (byteArray: Uint8Array, index: number) {
    // little endian
    let newInt = (
        ((0xFF & byteArray[index]) << 16) |
        ((0xFF & byteArray[index + 1]) << 8) |
        (0xFF & byteArray[index + 2])
    );
    if ((newInt & 0x00800000) > 0) {
        newInt |= 0xFF000000;
    } else {
        newInt &= 0x00FFFFFF;
    }
    return newInt;
}
